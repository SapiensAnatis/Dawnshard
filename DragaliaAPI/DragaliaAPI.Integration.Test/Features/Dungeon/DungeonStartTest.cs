using DragaliaAPI.Services.Game;
using DragaliaAPI.Shared.MasterAsset;
using DragaliaAPI.Shared.MasterAsset.Models;
using Microsoft.EntityFrameworkCore;
using Snapshooter;
using Snapshooter.Xunit;

namespace DragaliaAPI.Integration.Test.Features.Dungeon;

/// <summary>
/// Tests <see cref="DragaliaAPI.Features.Dungeon.Start.DungeonStartController"/>.
/// </summary>
public class DungeonStartTest : TestFixture
{
    public DungeonStartTest(CustomWebApplicationFactory factory, ITestOutputHelper outputHelper)
        : base(factory, outputHelper)
    {
        this.ImportSave().Wait();
    }

    [Fact]
    public async Task Start_OneTeam_HasExpectedPartyUnitList()
    {
        DungeonStartStartResponse response = (
            await Client.PostMsgpack<DungeonStartStartResponse>(
                "/dungeon_start/start",
                new DungeonStartStartRequest()
                {
                    PartyNoList = new List<int>() { 1 },
                    QuestId = 100010103
                }
            )
        ).Data;

        Snapshot.Match(response.IngameData.PartyInfo.PartyUnitList, SnapshotOptions);

        response.IngameData.PartyInfo.PartyUnitList.Should().HaveCount(4);
        response.IngameData.PartyInfo.PartyUnitList.Should().BeInAscendingOrder(x => x.Position);
        response.IngameData.PartyInfo.PartyUnitList.Should().OnlyHaveUniqueItems(x => x.Position);
        response.IngameData.IsBotTutorial.Should().BeFalse();
    }

    [Fact]
    public async Task Start_TwoTeams_HasExpectedPartyUnitList()
    {
        DungeonStartStartResponse response = (
            await Client.PostMsgpack<DungeonStartStartResponse>(
                "/dungeon_start/start",
                new DungeonStartStartRequest()
                {
                    PartyNoList = new List<int>() { 37, 38 },
                    QuestId = 100010103
                }
            )
        ).Data;

        // Abuse of snapshots here is lazy, but the resulting JSON is thousands of lines long...
        Snapshot.Match(response.IngameData.PartyInfo.PartyUnitList, SnapshotOptions);

        response.IngameData.PartyInfo.PartyUnitList.Should().HaveCount(8);
        response.IngameData.PartyInfo.PartyUnitList.Should().BeInAscendingOrder(x => x.Position);
        response.IngameData.PartyInfo.PartyUnitList.Should().OnlyHaveUniqueItems(x => x.Position);
    }

    [Fact]
    public async Task Start_WeaponPassivesUnlocked_IncludedInPartyUnitList()
    {
        DungeonStartStartResponse response = (
            await Client.PostMsgpack<DungeonStartStartResponse>(
                "/dungeon_start/start",
                new DungeonStartStartRequest()
                {
                    PartyNoList = new List<int>() { 38 },
                    QuestId = 100010103
                }
            )
        ).Data;

        response
            .IngameData.PartyInfo.PartyUnitList.First(x =>
                x.CharaData!.CharaId == Charas.GalaMascula
            )
            .GameWeaponPassiveAbilityList.Should()
            .Contain(x => x.WeaponPassiveAbilityId == 1020211);
    }

    [Fact]
    public async Task StartAssignUnit_HasExpectedPartyList()
    {
        DungeonSkipStartAssignUnitRequest request =
            new()
            {
                QuestId = 100010103,
                RequestPartySettingList = new List<PartySettingList>()
                {
                    new()
                    {
                        UnitNo = 1,
                        CharaId = Charas.GalaLeonidas,
                        EquipWeaponBodyId = WeaponBodies.Draupnir,
                        EquipDragonKeyId = (ulong)GetDragonKeyId(Dragons.Horus),
                        EquipCrestSlotType1CrestId1 = AbilityCrests.PrimalCrisis,
                        EquipCrestSlotType1CrestId2 = AbilityCrests.TheCutieCompetition,
                        EquipCrestSlotType1CrestId3 = AbilityCrests.AnIndelibleDate,
                        EquipCrestSlotType2CrestId1 = AbilityCrests.BeautifulGunman,
                        EquipCrestSlotType2CrestId2 = AbilityCrests.DragonArcanum,
                        EquipTalismanKeyId = (ulong)GetTalismanKeyId(Talismans.GalaLeonidas),
                        EquipCrestSlotType3CrestId1 = AbilityCrests.AKnightsDreamAxesBoon,
                        EquipCrestSlotType3CrestId2 = AbilityCrests.CrownofLightSerpentsBoon,
                        EditSkill1CharaId = Charas.GalaZethia,
                        EditSkill2CharaId = Charas.GalaMascula,
                    },
                    new()
                    {
                        UnitNo = 2,
                        CharaId = Charas.GalaGatov,
                        EquipWeaponBodyId = WeaponBodies.Mjoelnir,
                        EquipDragonKeyId = (ulong)GetDragonKeyId(Dragons.GalaMars),
                        EquipCrestSlotType1CrestId1 = AbilityCrests.TheCutieCompetition,
                        EquipCrestSlotType1CrestId2 = AbilityCrests.KungFuMasters,
                        EquipCrestSlotType1CrestId3 = AbilityCrests.BondsBetweenWorlds,
                        EquipCrestSlotType2CrestId1 = AbilityCrests.DragonArcanum,
                        EquipCrestSlotType2CrestId2 = AbilityCrests.BeautifulNothingness,
                        EquipTalismanKeyId = (ulong)GetTalismanKeyId(Talismans.GalaMym),
                        EquipCrestSlotType3CrestId1 = AbilityCrests.TutelarysDestinyWolfsBoon,
                        EquipCrestSlotType3CrestId2 = AbilityCrests.TestamentofEternityFishsBoon,
                    }
                }
            };

        DungeonStartStartAssignUnitResponse response = (
            await Client.PostMsgpack<DungeonStartStartAssignUnitResponse>(
                "/dungeon_start/start_assign_unit",
                request
            )
        ).Data;

        // Only test the first two since the others are empty
        Snapshot.Match(response.IngameData.PartyInfo.PartyUnitList.Take(2), SnapshotOptions);

        response.IngameData.PartyInfo.PartyUnitList.Should().HaveCount(4);
        response
            .IngameData.PartyInfo.PartyUnitList.Should()
            .Contain(x => x.CharaData!.CharaId == Charas.GalaLeonidas)
            .And.Contain(x => x.CharaData!.CharaId == Charas.GalaGatov);
    }

    [Theory]
    [InlineData("start")]
    [InlineData("start_assign_unit")]
    public async Task Start_InsufficientStamina_ReturnsError(string endpoint)
    {
        await this.ApiContext.PlayerUserData.ExecuteUpdateAsync(p =>
            p.SetProperty(e => e.StaminaSingle, e => 0)
        );
        await this.ApiContext.PlayerUserData.ExecuteUpdateAsync(p =>
            p.SetProperty(e => e.StaminaMulti, e => 0)
        );
        await this.ApiContext.PlayerUserData.ExecuteUpdateAsync(p =>
            p.SetProperty(e => e.LastStaminaSingleUpdateTime, e => DateTimeOffset.UtcNow)
        );
        await this.ApiContext.PlayerUserData.ExecuteUpdateAsync(p =>
            p.SetProperty(e => e.LastStaminaMultiUpdateTime, e => DateTimeOffset.UtcNow)
        );

        (
            await Client.PostMsgpack<DungeonStartStartResponse>(
                $"/dungeon_start/{endpoint}",
                new DungeonStartStartRequest() { QuestId = 100010104, PartyNoList = [1] },
                ensureSuccessHeader: false
            )
        )
            .DataHeaders.ResultCode.Should()
            .Be(ResultCode.QuestStaminaSingleShort);
    }

    [Fact]
    public async Task Start_ZeroStamina_FirstClearOfMainStory_Allows()
    {
        await this.ApiContext.PlayerQuests.ExecuteDeleteAsync();

        await this.ApiContext.PlayerUserData.ExecuteUpdateAsync(p =>
            p.SetProperty(e => e.StaminaSingle, e => 0)
        );
        await this.ApiContext.PlayerUserData.ExecuteUpdateAsync(p =>
            p.SetProperty(e => e.StaminaMulti, e => 0)
        );
        await this.ApiContext.PlayerUserData.ExecuteUpdateAsync(p =>
            p.SetProperty(e => e.LastStaminaSingleUpdateTime, e => DateTimeOffset.UtcNow)
        );
        await this.ApiContext.PlayerUserData.ExecuteUpdateAsync(p =>
            p.SetProperty(e => e.LastStaminaMultiUpdateTime, e => DateTimeOffset.UtcNow)
        );

        (
            await Client.PostMsgpack<DungeonStartStartResponse>(
                $"/dungeon_start/start",
                new DungeonStartStartRequest()
                {
                    QuestId = 100260101,
                    PartyNoList = new List<int>() { 1 },
                },
                ensureSuccessHeader: false
            )
        ).DataHeaders.ResultCode.Should().Be(ResultCode.Success);
    }

    [Fact]
    public async Task Start_ChronosClash_HasRareEnemy()
    {
        DragaliaResponse<DungeonStartStartResponse> response =
            await this.Client.PostMsgpack<DungeonStartStartResponse>(
                $"/dungeon_start/start",
                new DungeonStartStartRequest() { QuestId = 204270302, PartyNoList = [1] }
            );

        response.Data.OddsInfo.Enemy.Should().Contain(x => x.ParamId == 204130320 && x.IsRare);
    }

    [Fact]
    public async Task Start_EarnEvent_EnemiesDuplicated()
    {
        const int earnEventQuestId = 229031201; // Repelling the Frosty Fiends: Standard (Solo)

        DragaliaResponse<DungeonStartStartResponse> response =
            await this.Client.PostMsgpack<DungeonStartStartResponse>(
                $"/dungeon_start/start",
                new DungeonStartStartRequest() { QuestId = earnEventQuestId, PartyNoList = [1] }
            );

        response.Data.OddsInfo.Enemy.Should().HaveCount(31);

        QuestData questData = MasterAsset.QuestData[earnEventQuestId];
        IEnumerable<int> enemies = MasterAsset
            .QuestEnemies[$"{questData.Scene01}/{questData.AreaName01}".ToLowerInvariant()]
            .Enemies[questData.VariationType];

        response.Data.OddsInfo.Enemy.Should().HaveCountGreaterThan(enemies.Count());
    }

    [Fact]
    public async Task Start_CoopTutorial_SetsIsBotTutorial()
    {
        await this
            .ApiContext.PlayerUserData.Where(x => x.ViewerId == this.ViewerId)
            .ExecuteUpdateAsync(e =>
                e.SetProperty(p => p.TutorialStatus, TutorialService.TutorialStatusIds.CoopTutorial)
            );

        DragaliaResponse<DungeonStartStartResponse> response =
            await this.Client.PostMsgpack<DungeonStartStartResponse>(
                $"/dungeon_start/start",
                new DungeonStartStartRequest()
                {
                    QuestId = TutorialService.TutorialQuestIds.AvenueToPowerBeginner,
                    PartyNoList = [1]
                }
            );

        response.Data.IngameData.IsBotTutorial.Should().BeTrue();
    }

    [Fact]
    public async Task Start_AtpBeginner_NotCoopTutorial_SetsIsBotTutorial()
    {
        await this
            .ApiContext.PlayerUserData.Where(x => x.ViewerId == this.ViewerId)
            .ExecuteUpdateAsync(e =>
                e.SetProperty(
                    p => p.TutorialStatus,
                    TutorialService.TutorialStatusIds.CoopTutorial + 1
                )
            );

        DragaliaResponse<DungeonStartStartResponse> response =
            await this.Client.PostMsgpack<DungeonStartStartResponse>(
                $"/dungeon_start/start",
                new DungeonStartStartRequest()
                {
                    QuestId = TutorialService.TutorialQuestIds.AvenueToPowerBeginner,
                    PartyNoList = [1]
                }
            );

        response.Data.IngameData.IsBotTutorial.Should().BeFalse();
    }

    private static readonly Func<MatchOptions, MatchOptions> SnapshotOptions = opts =>
        opts.IgnoreField<long>("$..DragonData.DragonKeyId")
            .IgnoreField<long>("$..TalismanData.TalismanKeyId");
}
