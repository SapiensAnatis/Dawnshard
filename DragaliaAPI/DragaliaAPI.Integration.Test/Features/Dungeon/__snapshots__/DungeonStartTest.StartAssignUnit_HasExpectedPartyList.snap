[
  {
    "position": 1,
    "chara_data": {
      "chara_id": "GalaLeonidas",
      "exp": 1191950,
      "level": 80,
      "additional_max_level": 0,
      "hp": 754,
      "attack": 505,
      "ex_ability_level": 5,
      "ex_ability_2_level": 5,
      "ability_1_level": 2,
      "ability_2_level": 2,
      "ability_3_level": 2,
      "is_new": 1,
      "skill_1_level": 3,
      "skill_2_level": 2,
      "burst_attack_level": 2,
      "rarity": 5,
      "limit_break_count": 4,
      "hp_plus_count": 100,
      "attack_plus_count": 100,
      "status_plus_count": 0,
      "combo_buildup_count": 0,
      "is_unlock_edit_skill": false,
      "gettime": "2020-12-28T06:38:44+00:00",
      "mana_circle_piece_id_list": [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        21,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        31,
        32,
        33,
        34,
        35,
        36,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
        44,
        45,
        46,
        47,
        48,
        49,
        50
      ],
      "is_temporary": 0,
      "list_view_flag": 1
    },
    "dragon_data": {
      "dragon_id": "Horus",
      "dragon_key_id": 90,
      "level": 100,
      "exp": 1240020,
      "is_lock": true,
      "is_new": true,
      "get_time": "2020-10-05T08:39:40+00:00",
      "skill_1_level": 2,
      "ability_1_level": 5,
      "ability_2_level": 0,
      "limit_break_count": 4,
      "hp_plus_count": 50,
      "attack_plus_count": 50,
      "status_plus_count": 0
    },
    "weapon_skin_data": {
      "weapon_skin_id": 0
    },
    "weapon_body_data": {
      "weapon_body_id": "Draupnir",
      "buildup_count": 90,
      "limit_break_count": 9,
      "limit_over_count": 2,
      "skill_no": 1,
      "skill_level": 2,
      "ability_1_level": 3,
      "ability_2_level": 3,
      "equipable_count": 2,
      "additional_crest_slot_type_1_count": 1,
      "additional_crest_slot_type_2_count": 0,
      "additional_crest_slot_type_3_count": 2,
      "additional_effect_count": 0
    },
    "crest_slot_type_1_crest_list": [
      {
        "ability_crest_id": "PrimalCrisis",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "TheCutieCompetition",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "AnIndelibleDate",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 3,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_2_crest_list": [
      {
        "ability_crest_id": "BeautifulGunman",
        "buildup_count": 40,
        "limit_break_count": 4,
        "equipable_count": 2,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "DragonArcanum",
        "buildup_count": 10,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_3_crest_list": [
      {
        "ability_crest_id": "AKnightsDreamAxesBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      },
      {
        "ability_crest_id": "CrownofLightSerpentsBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 3,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      }
    ],
    "talisman_data": {
      "talisman_key_id": 21,
      "talisman_id": "GalaLeonidas",
      "talisman_ability_id_1": 340000049,
      "talisman_ability_id_2": 340000133,
      "talisman_ability_id_3": 0,
      "additional_hp": 0,
      "additional_attack": 0,
      "is_new": true,
      "is_lock": true,
      "gettime": "2021-12-30T00:55:01+00:00"
    },
    "edit_skill_1_chara_data": {
      "chara_id": "GalaZethia",
      "edit_skill_level": 3
    },
    "edit_skill_2_chara_data": {
      "chara_id": "GalaMascula",
      "edit_skill_level": 2
    },
    "game_weapon_passive_ability_list": [],
    "dragon_reliability_level": 30
  },
  {
    "position": 2,
    "chara_data": {
      "chara_id": "GalaGatov",
      "exp": 1191950,
      "level": 80,
      "additional_max_level": 0,
      "hp": 742,
      "attack": 518,
      "ex_ability_level": 5,
      "ex_ability_2_level": 5,
      "ability_1_level": 2,
      "ability_2_level": 2,
      "ability_3_level": 2,
      "is_new": 1,
      "skill_1_level": 3,
      "skill_2_level": 2,
      "burst_attack_level": 2,
      "rarity": 5,
      "limit_break_count": 4,
      "hp_plus_count": 100,
      "attack_plus_count": 100,
      "status_plus_count": 0,
      "combo_buildup_count": 0,
      "is_unlock_edit_skill": false,
      "gettime": "2022-04-05T07:31:10+00:00",
      "mana_circle_piece_id_list": [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        15,
        16,
        17,
        18,
        19,
        20,
        21,
        22,
        23,
        24,
        25,
        26,
        27,
        28,
        29,
        30,
        31,
        32,
        33,
        34,
        35,
        36,
        37,
        38,
        39,
        40,
        41,
        42,
        43,
        44,
        45,
        46,
        47,
        48,
        49,
        50
      ],
      "is_temporary": 0,
      "list_view_flag": 1
    },
    "dragon_data": {
      "dragon_id": "GalaMars",
      "dragon_key_id": 73,
      "level": 100,
      "exp": 1240020,
      "is_lock": true,
      "is_new": true,
      "get_time": "2020-04-27T05:22:45+00:00",
      "skill_1_level": 2,
      "ability_1_level": 5,
      "ability_2_level": 5,
      "limit_break_count": 4,
      "hp_plus_count": 50,
      "attack_plus_count": 50,
      "status_plus_count": 0
    },
    "weapon_skin_data": {
      "weapon_skin_id": 0
    },
    "weapon_body_data": {
      "weapon_body_id": "Mjoelnir",
      "buildup_count": 80,
      "limit_break_count": 8,
      "limit_over_count": 1,
      "skill_no": 1,
      "skill_level": 2,
      "ability_1_level": 2,
      "ability_2_level": 2,
      "equipable_count": 1,
      "additional_crest_slot_type_1_count": 1,
      "additional_crest_slot_type_2_count": 0,
      "additional_crest_slot_type_3_count": 2,
      "additional_effect_count": 0
    },
    "crest_slot_type_1_crest_list": [
      {
        "ability_crest_id": "TheCutieCompetition",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "KungFuMasters",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 1,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "BondsBetweenWorlds",
        "buildup_count": 50,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_2_crest_list": [
      {
        "ability_crest_id": "DragonArcanum",
        "buildup_count": 10,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      },
      {
        "ability_crest_id": "BeautifulNothingness",
        "buildup_count": 40,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 50,
        "attack_plus_count": 50
      }
    ],
    "crest_slot_type_3_crest_list": [
      {
        "ability_crest_id": "TutelarysDestinyWolfsBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 4,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      },
      {
        "ability_crest_id": "TestamentofEternityFishsBoon",
        "buildup_count": 30,
        "limit_break_count": 4,
        "equipable_count": 1,
        "ability_1_level": 3,
        "ability_2_level": 3,
        "hp_plus_count": 40,
        "attack_plus_count": 40
      }
    ],
    "talisman_data": {
      "talisman_key_id": 1,
      "talisman_id": "GalaMym",
      "talisman_ability_id_1": 340000029,
      "talisman_ability_id_2": 340000077,
      "talisman_ability_id_3": 0,
      "additional_hp": 0,
      "additional_attack": 0,
      "is_new": true,
      "is_lock": true,
      "gettime": "2021-12-27T10:29:11+00:00"
    },
    "edit_skill_1_chara_data": {
      "chara_id": "Empty",
      "edit_skill_level": 0
    },
    "edit_skill_2_chara_data": {
      "chara_id": "Empty",
      "edit_skill_level": 0
    },
    "game_weapon_passive_ability_list": [],
    "dragon_reliability_level": 30
  }
]
